File: coll2tde.c
/* 
 * This file is part of coll2tde.
 *
 * coll2tde is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 3
 * as published by the Free Software Foundation;
 *
 * coll2tde is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with coll2tde; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 * 
 * ============================================================================
 * Copyright (C) 2014 Oz Nahum Tiram <nahumoz@gmail.com>
 * ============================================================================
 */

#include <stdio.h>
#include <getopt.h>
#include "json.h"
#include "mongo.h"
#include "tde.h"

#define FIELDS 999 
#define HELP 998 


void print_usage() {
    printf("Usage: coll2tde -h HOST -d DATABASE -c COLLECTION [-q QUERY][--fields "
           "FIELDS] [-a AGGREGATION] -f TDEFILE\n");
}

void get_options(int argc, char *argv[], char **host, char **database,
    char **collection_name, char **filename, char **fields, char **query, 
    char **aggregation) {

    int opt = 0;
    int long_index = 0;
    
    static struct option long_options[] = {
        {"host",      required_argument,       0,  'h' },
        {"database", required_argument,       0,  'd' },
        {"collection",    required_argument, 0,  'c' },
        {"fields",   required_argument, 0,  COLS },
        {"query",   required_argument, 0,  'q' },
        {"aggregation",   required_argument, 0,  'a' },
        {"filename",   required_argument, 0,  'o' },
        {"help",   no_argument, 0,  HELP },
        {0,           0,                 0,  0   }
    };

    while ((opt = getopt_long(argc, argv,"h:d:c:f:q:a:", 
                   long_options, &long_index )) != -1) {
        switch (opt) {
             
            case 'h' : *host = optarg;
                 break;
             case 'd' : *database = optarg;
                 break;
             case 'c' : *collection_name = optarg; 
                 break;
             case 'f' : *filename = optarg;
                 break;
             case 'q' : *query = optarg;
                 break;
             case FIELDS: *fields = optarg;
                 break;
             case 'a' : *aggregation = optarg;
                 break;
             default: print_usage(); 
                 exit(EXIT_FAILURE);
        }
    }
    
    if ( *host == NULL || *database == NULL || *collection_name == NULL)  {
        print_usage();
        exit(EXIT_FAILURE);
        }

    /* if no file name is given the collection_name will be appended tde */
    //*filename = (char*)realloc(*filename, strlen(*collection_name)*sizeof(char)+4);
    //if (strlen(*filename) == 0 ) {
    if (*filename == NULL ) {
       *filename = (char*)realloc(*filename, strlen(*collection_name)*sizeof(char)+5);
        strcpy(*filename, *collection_name);
        strcat(*filename, ".tde");  
    }
    
    if ( *query && *aggregation ) {
        printf("Aggregation and query are mutually exclusive...\n");
        exit(EXIT_FAILURE);
    }

    
}

int
main (int argc, char *argv[]){
    
    /* we could pack all these variables in a struct */
    char *host = NULL;
    char *database = NULL;
    char *collection_name = NULL;
    char *filename = NULL;
    char *fields = NULL;
    char *query = NULL;
    char *aggregation = NULL;
    
    get_options(argc, argv, &host, &database, &collection_name, &filename, 
            &fields, &query, &aggregation);
    
    mongoc_client_t *client_p;
    mongoc_collection_t *collection_p;
    mongoc_cursor_t *cursor = NULL;
    const bson_t *doc;
    TAB_HANDLE hExtract;
    TAB_HANDLE hTable;
    TAB_HANDLE hTableDef;
    wchar_t *fname_w = calloc(strlen(filename) + 1, sizeof(wchar_t));
    char *jsstr = NULL;
    TAB_TYPE *column_types = NULL;
    int ncol = 0;
    int r = 0;
    int bHasTable;
    
    mbstowcs(fname_w, filename, strlen(filename)+1);
    TableauWChar *sfname = calloc(strlen(filename) + 1, sizeof(TableauWChar));
    TableauWChar sExtract[8];
    ToTableauString(fname_w, sfname);
    ToTableauString(L"Extract", sExtract);
    TryOp(TabExtractCreate(&hExtract, sfname));
    
    free(sfname);
    free(filename); 

    TryOp(TabExtractHasTable(hExtract, sExtract, &bHasTable));
    cursor = get_cursor(host, database, collection_name, query, fields,  
                        aggregation, &collection_p, &client_p);
    
    mongoc_cursor_next (cursor, &doc);
    if (!bHasTable) {
        jsstr = bson_as_json (doc, NULL);
        printf("Creating tde file: %ls\n", fname_w);
        free(fname_w);
        /* Table does not exist; create it. */
        hTableDef = make_table_definition(jsstr, &column_types, &ncol);
        TryOp(TabExtractAddTable(hExtract, sExtract, hTableDef, &hTable));
        puts("Successfully added table to file");
    } else {
        printf("Found existing file!\n");
        /* Open an existing table to add more rows. */
        TryOp(TabExtractOpenTable(hExtract, sExtract, &hTable));
        /* Get table definition and column_types */
        TryOp(TabTableGetTableDefinition(hTable, &hTableDef));
        get_columns(&column_types, hTableDef, &ncol);
    }
    
    /* insert all the data to DataExtract file */
    wchar_t **column_values = malloc(ncol * sizeof(wchar_t*));

    do {
        jsstr = bson_as_json (doc, NULL);
        jsmntok_t *tokens = json_tokenise(jsstr);
        extract_values(column_values, jsstr, tokens, &ncol);
        insert_values(column_values, column_types, hTable, ncol);
        r++;
    } while (mongoc_cursor_next (cursor, &doc)) ;
    
    printf("Successfully inserted %d rows\n", r);
    TryOp(TabExtractClose(hExtract));
    free(column_values);
    mongoc_cursor_destroy(cursor);
    mongoc_collection_destroy(collection_p);
    if (client_p)
        mongoc_client_destroy(client_p);
    mongoc_cleanup();  

    return 0;
}

File: py/coll2tde.py
#!/usr/bin/env python2

#  This file is part of coll2tde.
#
#  coll2tde is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License, version 3
#  as published by the Free Software Foundation;
#
#  coll2tde is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with coll2tde; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
#  ============================================================================
#  Copyright (C) 2014 Oz Nahum Tiram <nahumoz@gmail.com>
#  ============================================================================

"""
coll2tde.py is a re-implementation of coll2tde written in C. The purpose of
this program is to help benchmarking the C program, and provide a test bed for
quickly prototyping new features.
"""

import json
import sys
import datetime
import argparse
import pymongo
import bson
import dataextract as tde

u = ("Usage: coll2tde -s SERVER -d DATABASE -c COLLECTION [-q QUERY][--fields "
     "FIELDS]"
     "[-a AGGREGATION] -f TDEFILE")


def parser():
    parser = argparse.ArgumentParser(usage=u)

    parser.add_argument("-s", "--server", help="server name", required=True)
    parser.add_argument("-d", "--database", help="database",
                        required=True)
    parser.add_argument("-c", "--collection", help="collection name",
                        required=True)
    parser.add_argument("-q", "--query", help="query", required=False)
    parser.add_argument("-a", "--aggregation", help="aggregation",
                        required=False)
    parser.add_argument("-f", "--fields", help="field names", required=False)

    return parser


class CollectionToTDE(object):

    def __init__(self, args):
        self.host = args.server
        self.db = args.database
        self.collection = args.collection
        self.args = args

    def open_tableau_file(self, fname):
        pass

    def get_cursor(self, host, database, collection, query, fields,
                   aggregation):

        mongurl = host
        try:
            client = pymongo.MongoClient(mongurl)
            print client
        except:
            print "Could not parse mongo url"
            sys.exit(2)

        db = client[database]
        collection = db[collection]

        if not aggregation:
            for k, v in {'query': query, 'fields': fields}.items():
                if v:
                    try:
                        k = json.loads(v)
                    except ValueError:
                        print "could not understand your %s ..." % k
                        sys.exit()

            cursor = collection.find(query, fields)
        else:
            cursor = collection.aggregate(aggregation)

        return cursor

    def make_table_definition(self, record, column_types, column_names,
                              fileHandle):

        tableDef = tde.TableDefinition()

        for k, v in record.iteritems():
            if isinstance(v, bson.objectid.ObjectId):
                continue
            elif isinstance(v, int):
                column_types.append(tde.Type.INTEGER)
                tableDef.addColumn(k, tde.Type.INTEGER)
            elif isinstance(v, float):
                column_types.append(tde.Type.DOUBLE)
                tableDef.addColumn(k, tde.Type.DOUBLE)
            elif isinstance(v, unicode):
                column_types.append(tde.Type.UNICODE_STRING)
                tableDef.addColumn(k, tde.Type.UNICODE_STRING)
            elif isinstance(v, bool):
                column_types.append(tde.Type.BOOLEAN)
                tableDef.addColumn(k, tde.Type.BOOLEAN)
            elif isinstance(v, datetime.datetime):
                column_types.append(tde.Type.DATETIME)
                tableDef.addColumn(k, tde.Type.DATETIME)
            else:
                print "Found unknown type! values is of %s" % v.__class__
                sys.exit()
            column_names.append(k)

        return tableDef

    def get_or_create_table(self, filename, record):

        fileHandle = tde.Extract(filename)
        column_names = []
        column_types = []

        if not fileHandle.hasTable('Extract'):
            # Table does not exist; create it
            tableDef = self.make_table_definition(record, column_types,
                                                  column_names, fileHandle)
            table = fileHandle.addTable('Extract', tableDef)
            print column_types, column_names
        else:
            # Open an existing table to add more rows
            table = fileHandle.openTable('Extract')

        tableDef = table.getTableDefinition()

        return table, tableDef, fileHandle, column_types

    def insert_row(self, table, tableDef, column_types, rec):

        row = tde.Row(tableDef)
        i = 0

        if isinstance(rec['_id'], bson.objectid.ObjectId):
            rec.pop('_id')

        for k, v in rec.items():
            t = column_types[i]
            if t == tde.Type.BOOLEAN:
                row.setBoolean(i, int(v))
            elif t == tde.Type.INTEGER:
                row.setInteger(i, v)
            elif t == tde.Type.DOUBLE:
                row.setDouble(i, v)
            elif t == tde.Type.UNICODE_STRING:
                row.setString(i, v)
            elif t == tde.Type.DATETIME:
                row.setDateTime(i, v.year, v.month, v.day, v.hour,
                                v.minute, v.second, int(v.microsecond/10000))
            elif not v:
                row.setNull(i)

            i += 1

        # such an ugly API ... if you try to insert row after calling
        # row.close() you'll get a segmentation fault
        table.insert(row)
        row.close()

    def run(self, args):
        cursor = self.get_cursor(self.host, self.db, self.collection,
                                 args.query,
                                 args.fields, args.aggregation)
        rec = cursor.next()
        table, tableDef, fileHandle, column_types = self.get_or_create_table(
            'test.tde', rec)

        self.insert_row(table, tableDef, column_types, rec)

if __name__ == '__main__':

    args = parser().parse_args()

    print args.aggregation
    if args.aggregation and (args.query or args.fields):
        print "Aggregation and Qurey are mutually exclusive ..."
        sys.exit(2)

    coll2tde = CollectionToTDE(args)
    coll2tde.run(args)
